src/treq/test/test_auth.py

```diff
@@ -139,3 +162,95 @@ def test_add_unknown_auth(self):
+        auth_header = self._auth.build_authentication_header(
+            b'/spam/eggs', b'GET', False,
+            b'b7f36bc385a662ed615f27bd9e94eecd',
+            b'me@dragons', qop=None,
+            algorithm=b'MD5'
+        )
+        self.assertEquals(
+            auth_header,
+            b'Digest username="spam", realm="me@dragons", ' +
+            b'nonce="b7f36bc385a662ed615f27bd9e94eecd", ' +
+            b'uri="/spam/eggs", ' +
+            b'response="fc05d17c55156b278132a52dc0dca526", algorithm="MD5"',
+        )
+
+    def test_build_authentication_header_md5_sess_no_cache(self):
+        auth_header = self._auth.build_authentication_header(
+            b'/spam/eggs?ham=bacon', b'GET', False,
+            b'b7f36bc385a662ed615f27bd9e94eecd',
+            b'me@dragons', qop='auth',
+            algorithm=b'MD5-SESS'
+        )
+        self.assertRegex(
+            auth_header,
+            b'Digest username="spam", realm="me@dragons", ' +
+            b'nonce="b7f36bc385a662ed615f27bd9e94eecd", ' +
+            b'uri="/spam/eggs\\?ham=bacon", ' +
+            b'response="([0-9a-f]{32})", ' +
+            b'algorithm="MD5-SESS", qop="auth", ' +
+            b'nc=00000001, cnonce="([0-9a-f]{16})"',
+        )
+
+    def test_build_authentication_header_sha_no_cache_no_qop(self):
+        auth_header = self._auth.build_authentication_header(
+            b'/spam/eggs', b'GET', False,
+            b'b7f36bc385a662ed615f27bd9e94eecd',
+            b'me@dragons', qop=None,
+            algorithm=b'SHA'
+        )
+
+        self.assertEquals(
+            auth_header,
+            b'Digest username="spam", realm="me@dragons", ' +
+            b'nonce="b7f36bc385a662ed615f27bd9e94eecd", ' +
+            b'uri="/spam/eggs", ' +
+            b'response="45420a4786287998bcb99dfde563c3a198109b31", ' +
+            b'algorithm="SHA"'
+        )
+
+    def test_build_authentication_header_sha512_cache(self):
+        # Emulate 1st request
+        self._auth.build_authentication_header(
+            b'/spam/eggs', b'GET', False,
+            b'b7f36bc385a662ed615f27bd9e94eecd',
+            b'me@dragons', qop='auth',
+            algorithm=b'SHA-512'
+        )
+        # Get header after cached request
+        auth_header = self._auth.build_authentication_header(
+            b'/spam/eggs', b'GET', True,
+            b'b7f36bc385a662ed615f27bd9e94eecd',
+            b'me@dragons', qop='auth',
+            algorithm=b'SHA-512'
+        )
+
+        # Make sure metadata was cached
+        self.assertTrue(self._auth.cached_metadata_for(b'GET', b'/spam/eggs'))
+
+        self.assertRegex(
+            auth_header,
+            b'Digest username="spam", realm="me@dragons", ' +
+            b'nonce="b7f36bc385a662ed615f27bd9e94eecd", ' +
+            b'uri="/spam/eggs", ' +
+            b'response="([0-9a-f]{128})", ' +
+            b'algorithm="SHA-512", qop="auth", ' +
+            b'nc=00000002, cnonce="([0-9a-f]+?)"',
+        )
```

Comment:
Sorry, I know the surrounding code doesn't set the greatest example, but new tests should really have docstrings.

Reply:
Docstrings added for these